# Real-Valued QM Experiment Calculations

In [2]:
import numpy as np
from matplotlib import pyplot as plt

## Theoretical calculation using (standard) complex representation<a id='cqm1'></a>

Here we outline the theoretical calculations for the experiment. We need vector representations for basis states

$$|0\rangle\rightarrow\left(\begin{array}{c} 1\\ 0\end{array}\right), \;\;\;\;\;\;\;\; |1\rangle \rightarrow \left(\begin{array}{c} 0\\ 1\end{array}\right)$$

In this (usual) basis, the Pauli operators take the standard representations

$$\hat{\sigma}^{x} = \left(\begin{array}{cc} 0 & 1 \\ 1 & 0\end{array}\right),\;\;\;\;\;\;
\hat{\sigma}^{y} = \left(\begin{array}{cc} 0 & -i \\ i & 0\end{array}\right),\;\;\;\;\;\;
\hat{\sigma}^{z} = \left(\begin{array}{cc} 1 & 0 \\ 0 & -1\end{array}\right).$$

In [3]:
up = np.array([1,0])
do = np.array([0,1])

sx = np.array([[0,1],[1,0]])
sy = np.array([[0,-1j],[1j,0]])
sz = np.array([[1,0],[0,-1]])
so = np.array([[1,0],[0,1]])

Alice uses operators

$$\hat{A}_{1} = \hat{\sigma}^{z},\;\;\;\;\;\; \hat{A}_{2} = \hat{\sigma}^{x},\;\;\;\;\;\; \hat{A}_{3} = \hat{\sigma}^{y},$$

Bob uses

$$\hat{C}_{1} = \frac{\hat{\sigma}^{z}+\hat{\sigma}^{x}}{\sqrt{2}},\;\;\;\;\;\; \hat{C}_{2} = \frac{\hat{\sigma}^{z}-\hat{\sigma}^{x}}{\sqrt{2}},\;\;\;\;\;\; \hat{C}_{3} = \frac{\hat{\sigma}^{z}+\hat{\sigma}^{y}}{\sqrt{2}},$$

$$\hat{C}_{4} = \frac{\hat{\sigma}^{z}-\hat{\sigma}^{y}}{\sqrt{2}},\;\;\;\;\;\; \hat{C}_{5} = \frac{\hat{\sigma}^{x}+\hat{\sigma}^{y}}{\sqrt{2}},\;\;\;\;\;\; \hat{C}_{6} = \frac{\hat{\sigma}^{x}-\hat{\sigma}^{y}}{\sqrt{2}},$$


In [4]:
x1 = sz
x2 = sx
x3 = sy

z1 = (1/np.sqrt(2))*(sz+sx)
z2 = (1/np.sqrt(2))*(sz-sx)
z3 = (1/np.sqrt(2))*(sz+sy)
z4 = (1/np.sqrt(2))*(sz-sy)
z5 = (1/np.sqrt(2))*(sx+sy)
z6 = (1/np.sqrt(2))*(sx-sy)

Next we need to represent the entangled Bell states indexed by Bob's two BSM measurement results 00, 01, 10, 11 as

$$|\Phi_{00}\rangle = \frac{1}{\sqrt{2}}\left(|00\rangle + |11\rangle\right)$$
$$|\Phi_{01}\rangle = \frac{1}{\sqrt{2}}\left(|00\rangle - |11\rangle\right)$$
$$|\Phi_{10}\rangle = \frac{1}{\sqrt{2}}\left(|01\rangle + |10\rangle\right)$$
$$|\Phi_{11}\rangle = \frac{1}{\sqrt{2}}\left(|01\rangle - |10\rangle\right)$$

The basic two-qubit basis state representations are obtained from the tensor product

$$|00\rangle = |0\rangle\otimes |0\rangle \rightarrow \left(\begin{array}{c} 1\\ 0\end{array}\right)\otimes \left(\begin{array}{c} 1\\ 0\end{array}\right) = \left(\begin{array}{c} 1\\ 0 \\ 0\\ 0\end{array}\right)$$
$$|01\rangle = |0\rangle\otimes |1\rangle \rightarrow \left(\begin{array}{c} 1\\ 0\end{array}\right)\otimes \left(\begin{array}{c} 0\\ 1\end{array}\right) = \left(\begin{array}{c} 0\\ 1 \\ 0\\ 0\end{array}\right)$$
$$|10\rangle = |1\rangle\otimes |0\rangle \rightarrow \left(\begin{array}{c} 0\\ 1\end{array}\right)\otimes \left(\begin{array}{c} 1\\ 0\end{array}\right) = \left(\begin{array}{c} 0\\ 0 \\ 1\\ 0\end{array}\right)$$
$$|11\rangle = |1\rangle\otimes |1\rangle \rightarrow \left(\begin{array}{c} 0\\ 1\end{array}\right)\otimes \left(\begin{array}{c} 0\\ 1\end{array}\right) =\left(\begin{array}{c} 0\\ 0 \\ 0\\ 1\end{array}\right)$$

In [5]:
basis00 = np.array([1,0,0,0])
basis01 = np.array([0,1,0,0])
basis10 = np.array([0,0,1,0])
basis11 = np.array([0,0,0,1])

ϕ00 = (1/np.sqrt(2))*(basis00+basis11)
ϕ01 = (1/np.sqrt(2))*(basis00-basis11)
ϕ10 = (1/np.sqrt(2))*(basis01+basis10)
ϕ11 = (1/np.sqrt(2))*(basis01-basis10)

The initial state consists of two independently prepared Bell states of the form $|\Psi\rangle = |\Phi_{00}\rangle\otimes|\Phi_{00}\rangle$

In [6]:
ψ = np.kron(ϕ00,ϕ00)

The experiment consists of Alice and Charlie performing measurements on their qubits using their operators (labeled $x$ and $z$, respectively), obtaining values $a,c=\pm1$, respectively. Bob takes one qubit from each pair and performs a Bell state measurement on the pair, obtaining $b\equiv b_{2}b_{1} \in\left\{00,01,10,11\right\}$. In each instance, Alice and Bob randomly select operators from their respective sets. After many repetitions, they reconstruct the probability distribution $P(abc|xz)$. From this, they compute the following correlation or "score"

$$\Gamma = \sum_{b}\mathcal{T}_{b},$$

where

$$\mathcal{T}_{b} = (-1)^{b_{2}}(S_{11}^{b}+S_{12}^{b})+(-1)^{b_{1}}(S_{21}^{b}-S_{22}^{b})+(-1)^{b_{2}}(S_{13}^{b}+S_{14}^{b})-(-1)^{b_{1}+b_{2}}(S_{33}^{b}-S_{34}^{b})+(-1)^{b_{1}}(S_{25}^{b}+S_{26}^{b})-(-1)^{b_{1}+b_{2}}(S_{35}^{b}-S_{36}^{b}),$$

$$S_{xz}^{b} = \sum_{a,c=\pm1}P(abc|xz)ac,$$

In order to compute $P(abc|xz)$, we imagine eigenvectors $|a\rangle$ and $|c\rangle$ corresponding to measurments $a$ and $c$, respectively. Given Bob's measurement $b$ corresponding to the Bell state $|\Phi_{b}\rangle$, the probability is given by the Born rule

$$P(abc|xz) = \left|\left[\langle a| \otimes \langle \Phi_{b} | \otimes \langle c |\right] |\Psi\rangle \right|^{2}$$

The remaining work is to organize the eigenvalues and eigenvectors for each of Alice's and Charlie's operators in order to compute the probabilities. 

Below, we use `numpy` to compute eigenvalues and corresponding eigenvectors for each operator. The sorting operation ensures that the positive eigenvalue is listed before the negative eigenvalue.

In [7]:
a,b = np.linalg.eig(x1)

idx = a.argsort()[::-1]   
a = a[idx]
b = b[:,idx]

ex1m = b[:,1]
ex1p = b[:,0]

a,b = np.linalg.eig(x2)

idx = a.argsort()[::-1]   
a = a[idx]
b = b[:,idx]

ex2m = b[:,1]
ex2p = b[:,0]

a,b = np.linalg.eig(x3)

idx = a.argsort()[::-1]   
a = a[idx]
b = b[:,idx]

ex3m = b[:,1]
ex3p = b[:,0]

a,b = np.linalg.eig(z1)

idx = a.argsort()[::-1]   
a = a[idx]
b = b[:,idx]

ez1m = b[:,1]
ez1p = b[:,0]

a,b = np.linalg.eig(z2)

idx = a.argsort()[::-1]   
a = a[idx]
b = b[:,idx]

ez2m = b[:,1]
ez2p = b[:,0]

a,b = np.linalg.eig(z3)

idx = a.argsort()[::-1]   
a = a[idx]
b = b[:,idx]

ez3m = b[:,1]
ez3p = b[:,0]

a,b = np.linalg.eig(z4)

idx = a.argsort()[::-1]   
a = a[idx]
b = b[:,idx]

ez4m = b[:,1]
ez4p = b[:,0]

a,b = np.linalg.eig(z5)

idx = a.argsort()[::-1]   
a = a[idx]
b = b[:,idx]

ez5m = b[:,1]
ez5p = b[:,0]

a,b = np.linalg.eig(z6)

idx = a.argsort()[::-1]   
a = a[idx]
b = b[:,idx]

ez6m = b[:,1]
ez6p = b[:,0]

The following function computes $S_{xy}^{b}$, which is effectively the expectation value $\langle\Psi|\hat{A}\otimes |\Phi_{b}\rangle \langle \Phi_{b}| \otimes \hat{C}|\Psi\rangle$ given a particular outcome of Bob's measurement, $b$. The operator $|\Phi_{b}\rangle\langle\phi_{b}|$ projects out all components of the state orthogonal to $|\Phi_{b}\rangle$, corresponding to the condition that the expectation value is taken under the condition that Bob measures the state to be $|\Phi_{b}\rangle$.

In [8]:
def S(ψ,ap,am,ϕ,cp,cm):
    Sab = np.abs(ψ @ np.kron(ap,np.kron(ϕ,cp)))**2 + np.abs(ψ @ np.kron(am,np.kron(ϕ,cm)))**2 
    Sab = Sab - np.abs(ψ @ np.kron(am,np.kron(ϕ,cp)))**2 - np.abs(ψ @ np.kron(ap,np.kron(ϕ,cm)))**2

    return Sab

We use this function to calculate the $S_{xy}^{b}$ for each $x$, $y$, and state $b$. For each of the twelve combinations $xz\in\left\{11,12,21,22,13,14,33,34,25,26,35,36\right\}$ there is one case for each of the four values for $b$.

In [9]:
S00b11 = S(ψ,ex1p,ex1m,ϕ00,ez1p,ez1m)
S01b11 = S(ψ,ex1p,ex1m,ϕ01,ez1p,ez1m)
S10b11 = S(ψ,ex1p,ex1m,ϕ10,ez1p,ez1m)
S11b11 = S(ψ,ex1p,ex1m,ϕ11,ez1p,ez1m)

S00b12 = S(ψ,ex1p,ex1m,ϕ00,ez2p,ez2m)
S01b12 = S(ψ,ex1p,ex1m,ϕ01,ez2p,ez2m)
S10b12 = S(ψ,ex1p,ex1m,ϕ10,ez2p,ez2m)
S11b12 = S(ψ,ex1p,ex1m,ϕ11,ez2p,ez2m)

S00b21 = S(ψ,ex2p,ex2m,ϕ00,ez1p,ez1m)
S01b21 = S(ψ,ex2p,ex2m,ϕ01,ez1p,ez1m)
S10b21 = S(ψ,ex2p,ex2m,ϕ10,ez1p,ez1m)
S11b21 = S(ψ,ex2p,ex2m,ϕ11,ez1p,ez1m)

S00b22 = S(ψ,ex2p,ex2m,ϕ00,ez2p,ez2m)
S01b22 = S(ψ,ex2p,ex2m,ϕ01,ez2p,ez2m)
S10b22 = S(ψ,ex2p,ex2m,ϕ10,ez2p,ez2m)
S11b22 = S(ψ,ex2p,ex2m,ϕ11,ez2p,ez2m)

S00b13 = S(ψ,ex1p,ex1m,ϕ00,ez3p,ez3m)
S01b13 = S(ψ,ex1p,ex1m,ϕ01,ez3p,ez3m)
S10b13 = S(ψ,ex1p,ex1m,ϕ10,ez3p,ez3m)
S11b13 = S(ψ,ex1p,ex1m,ϕ11,ez3p,ez3m)

S00b14 = S(ψ,ex1p,ex1m,ϕ00,ez4p,ez4m)
S01b14 = S(ψ,ex1p,ex1m,ϕ01,ez4p,ez4m)
S10b14 = S(ψ,ex1p,ex1m,ϕ10,ez4p,ez4m)
S11b14 = S(ψ,ex1p,ex1m,ϕ11,ez4p,ez4m)

S00b33 = S(ψ,ex3p,ex3m,ϕ00,ez3p,ez3m)
S01b33 = S(ψ,ex3p,ex3m,ϕ01,ez3p,ez3m)
S10b33 = S(ψ,ex3p,ex3m,ϕ10,ez3p,ez3m)
S11b33 = S(ψ,ex3p,ex3m,ϕ11,ez3p,ez3m)

S00b34 = S(ψ,ex3p,ex3m,ϕ00,ez4p,ez4m)
S01b34 = S(ψ,ex3p,ex3m,ϕ01,ez4p,ez4m)
S10b34 = S(ψ,ex3p,ex3m,ϕ10,ez4p,ez4m)
S11b34 = S(ψ,ex3p,ex3m,ϕ11,ez4p,ez4m)

S00b25 = S(ψ,ex2p,ex2m,ϕ00,ez5p,ez5m)
S01b25 = S(ψ,ex2p,ex2m,ϕ01,ez5p,ez5m)
S10b25 = S(ψ,ex2p,ex2m,ϕ10,ez5p,ez5m)
S11b25 = S(ψ,ex2p,ex2m,ϕ11,ez5p,ez5m)

S00b26 = S(ψ,ex2p,ex2m,ϕ00,ez6p,ez6m)
S01b26 = S(ψ,ex2p,ex2m,ϕ01,ez6p,ez6m)
S10b26 = S(ψ,ex2p,ex2m,ϕ10,ez6p,ez6m)
S11b26 = S(ψ,ex2p,ex2m,ϕ11,ez6p,ez6m)

S00b35 = S(ψ,ex3p,ex3m,ϕ00,ez5p,ez5m)
S01b35 = S(ψ,ex3p,ex3m,ϕ01,ez5p,ez5m)
S10b35 = S(ψ,ex3p,ex3m,ϕ10,ez5p,ez5m)
S11b35 = S(ψ,ex3p,ex3m,ϕ11,ez5p,ez5m)

S00b36 = S(ψ,ex3p,ex3m,ϕ00,ez6p,ez6m)
S01b36 = S(ψ,ex3p,ex3m,ϕ01,ez6p,ez6m)
S10b36 = S(ψ,ex3p,ex3m,ϕ10,ez6p,ez6m)
S11b36 = S(ψ,ex3p,ex3m,ϕ11,ez6p,ez6m)

Lastly, we compute the $\mathcal{T}_{b}$ and sum over $b$ to obtain $\Gamma$.

In [11]:
b1=0
b2=0
T00 = ((-1)**b2)*(S00b11+S00b12)+((-1)**b1)*(S00b21-S00b22)+((-1)**b2)*(S00b13+S00b14)
T00 = T00 -((-1)**(b1+b2))*(S00b33-S00b34)+((-1)**b1)*(S00b25+S00b26)-((-1)**(b1+b2))*(S00b35-S00b36)

b1=1
b2=0
T01 = ((-1)**b2)*(S01b11+S01b12)+((-1)**b1)*(S01b21-S01b22)+((-1)**b2)*(S01b13+S01b14)
T01 = T01 -((-1)**(b1+b2))*(S01b33-S01b34)+((-1)**b1)*(S01b25+S01b26)-((-1)**(b1+b2))*(S01b35-S01b36)

b1=0
b2=1
T10 = ((-1)**b2)*(S10b11+S10b12)+((-1)**b1)*(S10b21-S10b22)+((-1)**b2)*(S10b13+S10b14)
T10 = T10 -((-1)**(b1+b2))*(S10b33-S10b34)+((-1)**b1)*(S10b25+S10b26)-((-1)**(b1+b2))*(S10b35-S10b36)

b1=1
b2=1
T11 = ((-1)**b2)*(S11b11+S11b12)+((-1)**b1)*(S11b21-S11b22)+((-1)**b2)*(S11b13+S11b14)
T11 = T11 -((-1)**(b1+b2))*(S11b33-S11b34)+((-1)**b1)*(S11b25+S11b26)-((-1)**(b1+b2))*(S11b35-S11b36)

print(T00+T01+T10+T11)

8.485281374238564


This value of $\Gamma$ can be compared to the theoretical value $6\sqrt{2}$.

In [12]:
6*np.sqrt(2)

8.485281374238571

## Theoretical calculation using real representation<a id='rqm1'></a>

One possible real representation which respects the usual tensor-product structure of multiparticle states is obtained by using a real, four-component vector to represent each qubit,
$$\left(\begin{array}{c} \alpha\\ \beta\end{array}\right)\rightarrow \left(\begin{array}\mbox{Re}[\alpha]\\ \mbox{Im}[\alpha]\\ \mbox{Re}[\beta]\\ \mbox{Im}[\beta]\end{array}\right).$$

One would like to generalize the representations,

$$|0\rangle \rightarrow \left(\begin{array}{c} 1 \\ 0 \\ 0 \\ 0\end{array}\right),\;\;\;\;\;\;\;\; |1\rangle \rightarrow \left(\begin{array}{c} 0 \\ 0 \\ 1 \\ 0\end{array}\right).$$

But now there are two <i>additional</i> basis states. What to make of these?

In the complex representation, the overall phase $e^{i\phi}$ is not measurable. That is, $i|0\rangle$ is indistinguishable from $|0\rangle$. This redundancy takes a slightly different form in the real representation where these additional basis states correspond to $i|0\rangle$, $i|1\rangle$. Let us define

$$|\overline{0}\rangle \rightarrow \left(\begin{array}{c} 0 \\ 1 \\ 0 \\ 0\end{array}\right),\;\;\;\;\;\;\;\; |\overline{1}\rangle \rightarrow \left(\begin{array}{c} 0 \\ 0 \\ 0 \\ 1\end{array}\right).$$

The usual phase invariance $e^{i\phi}|0\rangle$ can be captured in the real representation as $\cos\phi |0\rangle + \sin\phi |\overline{0}\rangle$ (and similarly, $e^{i\theta}|1\rangle\rightarrow \cos\theta|1\rangle + \sin\theta|\overline{1}\rangle$).

In this real basis, the spin operators take the representations

$$\hat{\sigma}^{x}_{r} = \left(\begin{array}{cccc} 0 & 0 & 0 & 1 \\0 & 0 & 1 & 0\\ 0 & 1 & 0 & 0 \\ 1 & 0 & 0 & 0\end{array}\right),\;\;\;\;\;\;
\hat{\sigma}^{y}_{r} = \left(\begin{array}{cccc} 0 & 0 & 0 & 1 \\ 0 & 0 & -1 & 0\\ 0 & -1 & 0 & 0 \\ 1 & 0 & 0 & 0\end{array}\right),\;\;\;\;\;\;
\hat{\sigma}^{z}_{r} = \left(\begin{array}{cccc} 1 & 0 & 0 & 0\\ 0 & 1 & 0 & 0 \\ 0 & 0 & -1 & 0 \\ 0 & 0 & 0 & -1\end{array}\right).$$

For reference, the real matrix representation of the imaginary unit is

$$i_{r}  = \left(\begin{array}{cccc} 0 & -1 & 0 & 0 \\ 1 & 0 & 0 & 0\\ 0 & 0 & 0 & -1 \\ 0 & 0 & 1 & 0\end{array}\right)$$

In [13]:
up = np.array([1,0,0,0])
do = np.array([0,0,1,0])
up1 = np.array([0,1,0,0])
do1 = np.array([0,0,0,1])

sx = np.array([[0,0,1,0],[0,0,0,1],[1,0,0,0],[0,1,0,0]])
sy = np.array([[0,0,0,1],[0,0,-1,0],[0,-1,0,0],[1,0,0,0]])
sz = np.array([[1,0,0,0],[0,1,0,0],[0,0,-1,0],[0,0,0,-1]])

ir = np.array([[0,-1,0,0],[1,0,0,0],[0,0,0,-1],[0,0,1,0]])

As in the complex representation, Alice uses operators

$$\hat{A}_{1} = \hat{\sigma}^{z},\;\;\;\;\;\; \hat{A}_{2} = \hat{\sigma}^{x},\;\;\;\;\;\; \hat{A}_{3} = \hat{\sigma}^{y},$$

Bob uses

$$\hat{C}_{1} = \frac{\hat{\sigma}^{z}+\hat{\sigma}^{x}}{\sqrt{2}},\;\;\;\;\;\; \hat{C}_{2} = \frac{\hat{\sigma}^{z}-\hat{\sigma}^{x}}{\sqrt{2}},\;\;\;\;\;\; \hat{C}_{3} = \frac{\hat{\sigma}^{z}+\hat{\sigma}^{y}}{\sqrt{2}},$$

$$\hat{C}_{4} = \frac{\hat{\sigma}^{z}-\hat{\sigma}^{y}}{\sqrt{2}},\;\;\;\;\;\; \hat{C}_{5} = \frac{\hat{\sigma}^{x}+\hat{\sigma}^{y}}{\sqrt{2}},\;\;\;\;\;\; \hat{C}_{6} = \frac{\hat{\sigma}^{x}-\hat{\sigma}^{y}}{\sqrt{2}},$$

The operator definitions are unchanged from the complex case but now use the real representations of the spin operators.

In [14]:
x1 = sz
x2 = sx
x3 = sy

z1 = (1/np.sqrt(2))*(sz+sx)
z2 = (1/np.sqrt(2))*(sz-sx)
z3 = (1/np.sqrt(2))*(sz+sy)
z4 = (1/np.sqrt(2))*(sz-sy)
z5 = (1/np.sqrt(2))*(sx+sy)
z6 = (1/np.sqrt(2))*(sx-sy)

Next we need to represent the entangled Bell states indexed by Bob's two BSM measurement results 00, 01, 10, 11 as

$$|\Phi_{00}\rangle = \frac{1}{\sqrt{2}}\left(|00\rangle + |11\rangle\right)$$
$$|\Phi_{01}\rangle = \frac{1}{\sqrt{2}}\left(|00\rangle - |11\rangle\right)$$
$$|\Phi_{10}\rangle = \frac{1}{\sqrt{2}}\left(|01\rangle + |10\rangle\right)$$
$$|\Phi_{11}\rangle = \frac{1}{\sqrt{2}}\left(|01\rangle - |10\rangle\right)$$

The new basis being $|0\rangle$, $1\rangle$, $|\overline{0}\rangle$, $|\overline{1}\rangle$ adds a bit of complexity. Formally, the two-particle basis becomes (take a deep breath) $|00\rangle$, $|0\overline{0}\rangle$, $|01\rangle$, $|\overline{0}\overline{1}\rangle$, $|\overline{0}0\rangle$, $|\overline{0}\overline{0}\rangle$, $|\overline{0}1\rangle$, $|1\overline{1}\rangle$, $|10\rangle$, $|1\overline{0}\rangle$, $|11\rangle$, $|1\overline{1}\rangle$, $|\overline{1}0\rangle$, $|\overline{1}\overline{0}\rangle$, $|\overline{1}1\rangle$, $|\overline{1}\overline{1}\rangle$.

The phase invariance results in four copies of each state. For example

$$|\Phi_{00}^{a}\rangle =\frac{1}{\sqrt{2}}\left(|00\rangle + |11\rangle\right)$$ 
$$|\Phi_{00}^{b}\rangle =\frac{1}{\sqrt{2}}\left(|\overline{0}0\rangle + |\overline{1}1\rangle\right)$$ 
$$|\Phi_{00}^{c}\rangle =\frac{1}{\sqrt{2}}\left(|0\overline{0}\rangle + |1\overline{1}\rangle\right)$$ 
$$|\Phi_{00}^{d}\rangle =\frac{1}{\sqrt{2}}\left(|\overline{0}\overline{0}\rangle + |\overline{1}\overline{1}\rangle\right)$$

All four states are mutually orthogonal, and similar definitions exist for the other Bell states.

In [15]:
ϕ00a = (1/np.sqrt(2))*(np.kron(up,up)+np.kron(do,do))
ϕ00b = (1/np.sqrt(2))*(np.kron(up1,up)+np.kron(do1,do))
ϕ00c = (1/np.sqrt(2))*(np.kron(up,up1)+np.kron(do,do1))
ϕ00d = (1/np.sqrt(2))*(np.kron(up1,up1)+np.kron(do1,do1))

ϕ01a = (1/np.sqrt(2))*(np.kron(up,up)-np.kron(do,do))
ϕ01b = (1/np.sqrt(2))*(np.kron(up1,up)-np.kron(do1,do))
ϕ01c = (1/np.sqrt(2))*(np.kron(up,up1)-np.kron(do,do1))
ϕ01d = (1/np.sqrt(2))*(np.kron(up1,up1)-np.kron(do1,do1))

ϕ10a = (1/np.sqrt(2))*(np.kron(up,do)+np.kron(do,up))
ϕ10b = (1/np.sqrt(2))*(np.kron(up1,do)+np.kron(do1,up))
ϕ10c = (1/np.sqrt(2))*(np.kron(up,do1)+np.kron(do,up1))
ϕ10d = (1/np.sqrt(2))*(np.kron(up1,do1)+np.kron(do1,up1))

ϕ11a = (1/np.sqrt(2))*(np.kron(up,do)-np.kron(do,up))
ϕ11b = (1/np.sqrt(2))*(np.kron(up1,do)-np.kron(do1,up))
ϕ11c = (1/np.sqrt(2))*(np.kron(up,do1)-np.kron(do,up1))
ϕ11d = (1/np.sqrt(2))*(np.kron(up1,do1)-np.kron(do1,up1))


For convenience later, each set of "redundant" states is packaged into a matrix with each row corresponding to one state.

In [16]:
ϕ00 = np.vstack((ϕ00a,ϕ00b,ϕ00c,ϕ00d))
ϕ01 = np.vstack((ϕ01a,ϕ01b,ϕ01c,ϕ01d))
ϕ10 = np.vstack((ϕ10a,ϕ10b,ϕ10c,ϕ10d))
ϕ11 = np.vstack((ϕ11a,ϕ11b,ϕ11c,ϕ11d))

The state $|\Psi\rangle = |\Phi_{00}\rangle\otimes |\Phi_{00}\rangle$ is defined as before using $|\Phi_{00}{a}\rangle$. The user is encouraged to try other states (and combinations of the Bell states). 

In [17]:
ψ = np.kron(ϕ00a,ϕ00a)

#ψ = (np.kron(ϕ00a,ϕ00a)+kron(ϕ00d,ϕ00d))/sqrt(2)
#ψ = (np.kron((ϕ00a+ϕ00b)/np.sqrt(2),(ϕ00c+ϕ00d)/np.sqrt(2)))

The experiment, as before, consists of Alice and Charlie performing measurements on their qubits using their operators (labeled $x$ and $z$, respectively), obtaining values $a,c=\pm1$, respectively. Bob takes one qubit from each pair and performs a Bell state measurement on the pair, obtaining $b\equiv b_{2}b_{1} \in\left\{00,01,10,11\right\}$. In each instance, Alice and Bob randomly select operators from their respective sets. After many repetitions, they reconstruct the probability distribution $P(abc|xz)$. From this, they compute the following correlation or "score"

$$\Gamma = \sum_{b}\mathcal{T}_{b},$$

where

$$\mathcal{T}_{b} = (-1)^{b_{2}}(S_{11}^{b}+S_{12}^{b})+(-1)^{b_{1}}(S_{21}^{b}-S_{22}^{b})+(-1)^{b_{2}}(S_{13}^{b}+S_{14}^{b})-(-1)^{b_{1}+b_{2}}(S_{33}^{b}-S_{34}^{b})+(-1)^{b_{1}}(S_{25}^{b}+S_{26}^{b})-(-1)^{b_{1}+b_{2}}(S_{35}^{b}-S_{36}^{b}),$$

$$S_{xz}^{b} = \sum_{a,c=\pm1}P(abc|xz)ac,$$

In order to compute $P(abc|xz)$, we imagine eigenvectors $|a\rangle$ and $|c\rangle$ corresponding to measurments $a$ and $c$, respectively. Given Bob's measurement $b$ corresponding to the Bell state $|\Phi_{b}\rangle$, the probability is given by the Born rule

$$P(abc|xz) = \left|\left[\langle a| \otimes \langle \Phi_{b} | \otimes \langle c |\right] |\Psi\rangle \right|^{2}$$

The remaining work is to organize the eigenvalues and eigenvectors for each of Alice's and Charlie's operators in order to compute the probabilities. 

Below, we use `numpy` to compute eigenvalues and corresponding eigenvectors for each operator. The sorting operation ensures that the positive eigenvalue is listed before the negative eigenvalue.

In the real representation, there is also a redundancy of operator eigenvalues (e.g., two positive eigenvalues and two negative eigenvalues). We need to keep track of all eigenvectors to properly compute measurement probabilities.

In [18]:
a,b = np.linalg.eigh(x1)

idx = a.argsort()[::-1]   
a = a[idx]
b = b[:,idx]

ex1m = b[:,2]
ex1p = b[:,0]
ex1m1 = b[:,3]
ex1p1 = b[:,1]

ex1_p = np.vstack((ex1p,ex1p1))
ex1_m = np.vstack((ex1m,ex1m1))

a,b = np.linalg.eigh(x2)

idx = a.argsort()[::-1]   
a = a[idx]
b = b[:,idx]

ex2m = b[:,2]
ex2p = b[:,0]
ex2m1 = b[:,3]
ex2p1 = b[:,1]

ex2_p = np.vstack((ex2p,ex2p1))
ex2_m = np.vstack((ex2m,ex2m1))

a,b = np.linalg.eigh(x3)

idx = a.argsort()[::-1]   
a = a[idx]
b = b[:,idx]

ex3m = b[:,2]
ex3p = b[:,0]
ex3m1 = b[:,3]
ex3p1 = b[:,1]

ex3_p = np.vstack((ex3p,ex3p1))
ex3_m = np.vstack((ex3m,ex3m1))

a,b = np.linalg.eigh(z1)

idx = a.argsort()[::-1]   
a = a[idx]
b = b[:,idx]

ez1m = b[:,2]
ez1p = b[:,0]
ez1m1 = b[:,3]
ez1p1 = b[:,1]

ez1_p = np.vstack((ez1p,ez1p1))
ez1_m = np.vstack((ez1m,ez1m1))

a,b = np.linalg.eigh(z2)

idx = a.argsort()[::-1]   
a = a[idx]
b = b[:,idx]

ez2m = b[:,2]
ez2p = b[:,0]
ez2m1 = b[:,3]
ez2p1 = b[:,1]

ez2_p = np.vstack((ez2p,ez2p1))
ez2_m = np.vstack((ez2m,ez2m1))

a,b = np.linalg.eigh(z3)

idx = a.argsort()[::-1]   
a = a[idx]
b = b[:,idx]

ez3m = b[:,2]
ez3p = b[:,0]
ez3m1 = b[:,3]
ez3p1 = b[:,1]

ez3_p = np.vstack((ez3p,ez3p1))
ez3_m = np.vstack((ez3m,ez3m1))

a,b = np.linalg.eigh(z4)

idx = a.argsort()[::-1]   
a = a[idx]
b = b[:,idx]

ez4m = b[:,2]
ez4p = b[:,0]
ez4m1 = b[:,3]
ez4p1 = b[:,1]

ez4_p = np.vstack((ez4p,ez4p1))
ez4_m = np.vstack((ez4m,ez4m1))

a,b = np.linalg.eigh(z5)

idx = a.argsort()[::-1]   
a = a[idx]
b = b[:,idx]

ez5m = b[:,2]
ez5p = b[:,0]
ez5m1 = b[:,3]
ez5p1 = b[:,1]

ez5_p = np.vstack((ez5p,ez5p1))
ez5_m = np.vstack((ez5m,ez5m1))

a,b = np.linalg.eigh(z6)

idx = a.argsort()[::-1]   
a = a[idx]
b = b[:,idx]

ez6m = b[:,2]
ez6p = b[:,0]
ez6m1 = b[:,3]
ez6p1 = b[:,1]

ez6_p = np.vstack((ez6p,ez6p1))
ez6_m = np.vstack((ez6m,ez6m1))

For a fixed measurement $(a,c)$, the following function takes a physically redundant set of eigenvectors $\left\{|a\rangle\right\}$ and a similar set of eigenvectors $\left\{|b\rangle\right\}$, the set of <i>four</i> physically redundant Bell states $|\Phi_{b}^{a,b,c,d}\rangle$, and the state $|\Psi\rangle$. Summing over the physical redundancies, we obtain the probability of obtaining $a,c$ under the assumption that Bob measures $|\Phi_{b}\rangle$. 

In [19]:
def probtot(ψ,a,ϕ,c):
    pt = 0

    for i in range(0,2):
        for j in range(0,4):
            for k in range(0,2):
                pt = pt + abs(ψ @ np.kron(a[i,:],np.kron(ϕ[j,:],c[k,:])))**2

    return pt

The following function sums over the configurations of $a,c$ and uses the function above to return the conditional expectation value $\langle\Psi|\hat{A}\otimes |\Phi_{b}\rangle \langle \Phi_{b}| \otimes \hat{C}|\Psi\rangle$ corresponding to $S_{xy}^{b}$.

In [20]:
def Sr(ψ,ap,am,ϕ,cp,cm):
    Sab = probtot(ψ,ap,ϕ,cp) + probtot(ψ,am,ϕ,cm) - probtot(ψ,ap,ϕ,cm) - probtot(ψ,am,ϕ,cp)
    return Sab

The following cell computes all $4\times 12=48$ cases of $S_{xy}^{b}$. The remainder of the computation of $\Gamma$ is identical in structure to the complex-valued case.

In [21]:
S00b11 = Sr(ψ,ex1_p,ex1_m,ϕ00,ez1_p,ez1_m)
S01b11 = Sr(ψ,ex1_p,ex1_m,ϕ01,ez1_p,ez1_m)
S10b11 = Sr(ψ,ex1_p,ex1_m,ϕ10,ez1_p,ez1_m)
S11b11 = Sr(ψ,ex1_p,ex1_m,ϕ11,ez1_p,ez1_m)

S00b12 = Sr(ψ,ex1_p,ex1_m,ϕ00,ez2_p,ez2_m)
S01b12 = Sr(ψ,ex1_p,ex1_m,ϕ01,ez2_p,ez2_m)
S10b12 = Sr(ψ,ex1_p,ex1_m,ϕ10,ez2_p,ez2_m)
S11b12 = Sr(ψ,ex1_p,ex1_m,ϕ11,ez2_p,ez2_m)

S00b21 = Sr(ψ,ex2_p,ex2_m,ϕ00,ez1_p,ez1_m)
S01b21 = Sr(ψ,ex2_p,ex2_m,ϕ01,ez1_p,ez1_m)
S10b21 = Sr(ψ,ex2_p,ex2_m,ϕ10,ez1_p,ez1_m)
S11b21 = Sr(ψ,ex2_p,ex2_m,ϕ11,ez1_p,ez1_m)

S00b22 = Sr(ψ,ex2_p,ex2_m,ϕ00,ez2_p,ez2_m)
S01b22 = Sr(ψ,ex2_p,ex2_m,ϕ01,ez2_p,ez2_m)
S10b22 = Sr(ψ,ex2_p,ex2_m,ϕ10,ez2_p,ez2_m)
S11b22 = Sr(ψ,ex2_p,ex2_m,ϕ11,ez2_p,ez2_m)

S00b13 = Sr(ψ,ex1_p,ex1_m,ϕ00,ez3_p,ez3_m)
S01b13 = Sr(ψ,ex1_p,ex1_m,ϕ01,ez3_p,ez3_m)
S10b13 = Sr(ψ,ex1_p,ex1_m,ϕ10,ez3_p,ez3_m)
S11b13 = Sr(ψ,ex1_p,ex1_m,ϕ11,ez3_p,ez3_m)

S00b14 = Sr(ψ,ex1_p,ex1_m,ϕ00,ez4_p,ez4_m)
S01b14 = Sr(ψ,ex1_p,ex1_m,ϕ01,ez4_p,ez4_m)
S10b14 = Sr(ψ,ex1_p,ex1_m,ϕ10,ez4_p,ez4_m)
S11b14 = Sr(ψ,ex1_p,ex1_m,ϕ11,ez4_p,ez4_m)

S00b33 = Sr(ψ,ex3_p,ex3_m,ϕ00,ez3_p,ez3_m)
S01b33 = Sr(ψ,ex3_p,ex3_m,ϕ01,ez3_p,ez3_m)
S10b33 = Sr(ψ,ex3_p,ex3_m,ϕ10,ez3_p,ez3_m)
S11b33 = Sr(ψ,ex3_p,ex3_m,ϕ11,ez3_p,ez3_m)

S00b34 = Sr(ψ,ex3_p,ex3_m,ϕ00,ez4_p,ez4_m)
S01b34 = Sr(ψ,ex3_p,ex3_m,ϕ01,ez4_p,ez4_m)
S10b34 = Sr(ψ,ex3_p,ex3_m,ϕ10,ez4_p,ez4_m)
S11b34 = Sr(ψ,ex3_p,ex3_m,ϕ11,ez4_p,ez4_m)

S00b25 = Sr(ψ,ex2_p,ex2_m,ϕ00,ez5_p,ez5_m)
S01b25 = Sr(ψ,ex2_p,ex2_m,ϕ01,ez5_p,ez5_m)
S10b25 = Sr(ψ,ex2_p,ex2_m,ϕ10,ez5_p,ez5_m)
S11b25 = Sr(ψ,ex2_p,ex2_m,ϕ11,ez5_p,ez5_m)

S00b26 = Sr(ψ,ex2_p,ex2_m,ϕ00,ez6_p,ez6_m)
S01b26 = Sr(ψ,ex2_p,ex2_m,ϕ01,ez6_p,ez6_m)
S10b26 = Sr(ψ,ex2_p,ex2_m,ϕ10,ez6_p,ez6_m)
S11b26 = Sr(ψ,ex2_p,ex2_m,ϕ11,ez6_p,ez6_m)

S00b35 = Sr(ψ,ex3_p,ex3_m,ϕ00,ez5_p,ez5_m)
S01b35 = Sr(ψ,ex3_p,ex3_m,ϕ01,ez5_p,ez5_m)
S10b35 = Sr(ψ,ex3_p,ex3_m,ϕ10,ez5_p,ez5_m)
S11b35 = Sr(ψ,ex3_p,ex3_m,ϕ11,ez5_p,ez5_m)

S00b36 = Sr(ψ,ex3_p,ex3_m,ϕ00,ez6_p,ez6_m)
S01b36 = Sr(ψ,ex3_p,ex3_m,ϕ01,ez6_p,ez6_m)
S10b36 = Sr(ψ,ex3_p,ex3_m,ϕ10,ez6_p,ez6_m)
S11b36 = Sr(ψ,ex3_p,ex3_m,ϕ11,ez6_p,ez6_m)

Given $S_{xy}^{b}$, the next step is to compute
$$\mathcal{T}_{b} = (-1)^{b_{2}}(S_{11}^{b}+S_{12}^{b})+(-1)^{b_{1}}(S_{21}^{b}-S_{22}^{b})+(-1)^{b_{2}}(S_{13}^{b}+S_{14}^{b})-(-1)^{b_{1}+b_{2}}(S_{33}^{b}-S_{34}^{b})+(-1)^{b_{1}}(S_{25}^{b}+S_{26}^{b})-(-1)^{b_{1}+b_{2}}(S_{35}^{b}-S_{36}^{b}).$$

The score is then given by 

$$\Gamma = \sum_{b}\mathcal{T}_{b}$$

In [22]:
b1=0
b2=0
T00 = ((-1)**b2)*(S00b11+S00b12)+((-1)**b1)*(S00b21-S00b22)+((-1)**b2)*(S00b13+S00b14)
T00 = T00 -((-1)**(b1+b2))*(S00b33-S00b34)+((-1)**b1)*(S00b25+S00b26)-((-1)**(b1+b2))*(S00b35-S00b36)

b1=1
b2=0
T01 = ((-1)**b2)*(S01b11+S01b12)+((-1)**b1)*(S01b21-S01b22)+((-1)**b2)*(S01b13+S01b14)
T01 = T01 -((-1)**(b1+b2))*(S01b33-S01b34)+((-1)**b1)*(S01b25+S01b26)-((-1)**(b1+b2))*(S01b35-S01b36)

b1=0
b2=1
T10 = ((-1)**b2)*(S10b11+S10b12)+((-1)**b1)*(S10b21-S10b22)+((-1)**b2)*(S10b13+S10b14)
T10 = T10 -((-1)**(b1+b2))*(S10b33-S10b34)+((-1)**b1)*(S10b25+S10b26)-((-1)**(b1+b2))*(S10b35-S10b36)

b1=1
b2=1
T11 = ((-1)**b2)*(S11b11+S11b12)+((-1)**b1)*(S11b21-S11b22)+((-1)**b2)*(S11b13+S11b14)
T11 = T11 -((-1)**(b1+b2))*(S11b33-S11b34)+((-1)**b1)*(S11b25+S11b26)-((-1)**(b1+b2))*(S11b35-S11b36)

print(T00+T01+T10+T11)

5.656854249492376


The result (for $|\Psi\rangle = |\Phi_{00}^{a}\rangle\otimes |\Phi_{00}^{a}\rangle$) corresponds to $4\sqrt{2}$.

In [23]:
4*np.sqrt(2)

5.656854249492381

# Expectation value formulation

## Complex-valued QM<a id='cqm2'></a>

As shown in the main text, the computation of $\Gamma$ can be reframed as an expectation value

$$\Gamma = \langle \Psi' | \mathcal{O} |\Psi'\rangle,$$

where $|\Psi'\rangle = \hat{H}_{1}\mbox{CNOT}_{1,2}|\Psi\rangle$ and

$$\mathcal{O} = 2\sqrt{2}\left[\hat{\sigma}^{x}\otimes\hat{\sigma}^{z}\otimes\hat{I}\otimes\hat{\sigma}^{x} - \hat{\sigma}^{y}\otimes\hat{\sigma}^{z}\otimes\hat{\sigma}^{z}\otimes\hat{\sigma}^{y} + \hat{\sigma}^{z}\otimes\hat{I}\otimes\hat{\sigma}^{z}\otimes\hat{\sigma}^{z}\right]$$

For a system of four qubits, the labeling convention is $0,1,2,3$ (top to bottom on circuit). This goes against Qiskit convention, but there are only a couple instances this will matter. The operation $\hat{H}_{1}$ denotes a Hadamard gate acting on qubit 1, and $\mbox{CNOT}_{1,2}$ denotes a controlled NOT operation with control on qubit 1 and target on qubit 2. A single Hadamard operation has the matrix representation

$$\hat{H} \;\dot{=}\; \frac{1}{\sqrt{2}}\left(\begin{array}{cc} 1 & 1\\ 1 & -1 \end{array}\right)$$

The CNOT operation (assuming basis $\left\{|00\rangle, |01\rangle, |10\rangle, |11\rangle\right\}$ in the form (control,target)) takes the matrix representation

$$\mbox{CNOT} \;\dot{=}\; \left(\begin{array}{cccc} 1 & 0 & 0 & 0 \\ 0 & 1 & 0 & 0 \\ 0 & 0 & 0 & 1\\ 0 & 0 & 1 & 0\end{array}\right).$$

This setup results in a much more efficient calculation of the theoretical result <i>and</i> a much more efficient circuit implementation using `Estimator`, but it does not result in a direct computation/measurement of the joint conditional probability distribution $P(abc|xz)$.

To proceed, we define the state $|\Psi'\rangle$ by applying the appropriate operations to $|\Phi\rangle = |\Phi_{00}\rangle\otimes|\Phi_{00}\rangle$, construct the matrix representation of $\mathcal{O}$, and compute the expectation value $\langle \Psi' |\mathcal{O}|\Psi'\rangle$. We use the shorthand `so` for the ($2\times 2$) identity matrix. The operation `@` performs matrix multiplication.


In [24]:
up = np.array([1,0])
do = np.array([0,1])

ϕ00 = (np.kron(up,up)+np.kron(do,do))/np.sqrt(2)

ψ = np.kron(ϕ00,ϕ00)

sz = np.array([[1,0],[0,-1]])
sx = np.array([[0,1],[1,0]])
sy = np.array([[0,-1j],[1j,0]])
so = np.array([[1,0],[0,1]]) #identity matrix

H = (1/np.sqrt(2))*np.array([[1,1],[1,-1]])
CNOT = np.array([[1,0,0,0],[0,1,0,0],[0,0,0,1],[0,0,1,0]])

ψp = np.kron(so,np.kron(H,np.kron(so,so))) @ np.kron(so,np.kron(CNOT,so)) @ ψ

O = 2*np.sqrt(2)*np.kron(sx,np.kron(sz,np.kron(so,sx)))
O = O - 2*np.sqrt(2)*np.kron(sy,np.kron(sz,np.kron(sz,sy)))
O = O + 2*np.sqrt(2)*np.kron(sz,np.kron(so,np.kron(sz,sz)))

print(ψp @ O @ ψp)

(8.485281374238566+0j)


## Real-valued QM <a id='rqm2'></a>

The same type of calculation can be performed using the real representation, so we include it for completion. 

The essential modifications are as follows.

1. The Hadamard operation acts as $\hat{H}|0\rangle= (|0\rangle + |1\rangle)/\sqrt{2}$ and $\hat{H}|1\rangle = (|0\rangle - |1\rangle)/\sqrt{2}$. Since $|\overline{0}\rangle$ corresponds to $i|0\rangle$ in the real representation (and similarly for $|\overline{1}\rangle$, we take $\hat{H}|\overline{0}\rangle= (|\overline{0}\rangle + |\overline{1}\rangle)/\sqrt{2}$ and $\hat{H}|\overline{1}\rangle = (|\overline{0}\rangle - |\overline{1}\rangle)/\sqrt{2}$. This leads to a real representation for the Hadamard operation as
$$\hat{H} \; \dot{=} \; \frac{1}{\sqrt{2}}\left(\begin{array}{cccc} 1 & 0 & 1 & 0 \\ 0 & 1 & 0 & 1\\ 1 & 0 & -1 & 0 \\ 0 & 1 & 0 & -1\end{array}\right)$$

2. The CNOT operation acts on two qubits. We take the operation defined in the complex representation (see above) and extend it to the real representation as with the Hadamard gate (e.g., $\mbox{CNOT}|1\mathcal{0}\rangle = |1\overline{1}\rangle$). This leads to a $16\times 16$ matrix of the form
$$\mbox{CNOT} = \left(\begin{array}{cc}{\bf I} & {\bf 0}\\ {\bf 0} & {\bf M}\end{array}\right)$$

$\;\;\;\;\;\;$Above, ${\bf I}$ is the eight-dimensional identity, ${\bf 0}$ is an $8\times 8$ matrix of zeros, and

$${\bf M} = \left(\begin{array}{cccccccc} 0 & 0 & 1 & 0 & 0 & 0 & 0 & 0 \\ 0 & 0 & 0 & 1 & 0 & 0 & 0 & 0 \\ 1 & 0 & 0 & 0 & 0 & 0 & 0 & 0 \\ 0 & 1 & 0 & 0 & 0 & 0 & 0 & 0 \\ 0 & 0 & 0 & 0 & 0 & 0 & 1 & 0 \\ 0 & 0 & 0 & 0 & 0 & 0 & 0 & 1 \\ 0 & 0 & 0 & 0 & 1 & 0 & 0 & 0 \\ 0 & 0 & 0 & 0 & 0 & 1 & 0 & 0\end{array}\right)$$

Aside from these changes (and the four-dimensional representation for the initial state and operators), the calculation proceeds identically to the complex case.

In [25]:
up = np.array([1,0,0,0])
do = np.array([0,0,1,0])
up1 = np.array([0,1,0,0])
do1 = np.array([0,0,0,1])

ϕ00a = (np.kron(up,up)+np.kron(do,do))/np.sqrt(2)
ϕ00d = (np.kron(up1,up1)+np.kron(do1,do1))/np.sqrt(2)

ψ = np.kron(ϕ00a,ϕ00a)

sz = np.array([[1,0,0,0],[0,1,0,0],[0,0,-1,0],[0,0,0,-1]])
sx = np.array([[0,0,1,0],[0,0,0,1],[1,0,0,0],[0,1,0,0]])
sy = np.array([[0,0,0,1],[0,0,-1,0],[0,-1,0,0],[1,0,0,0]])
so = np.eye(4)

H = (1/np.sqrt(2))*np.array([[1,0,1,0],[1,0,1,0],[0,1,0,-1],[0,1,0,-1]])

Z = np.zeros((8,8))
I = np.eye(8)
M = np.zeros((8,8))

M[0,2] = 1
M[1,3] = 1
M[2,0] = 1
M[3,1] = 1
M[4,6] = 1
M[5,7] = 1
M[6,4] = 1
M[7,5] = 1

CNOT = np.vstack((np.hstack((I,Z)),np.hstack((Z,M))))

ψp = np.kron(so,np.kron(H,np.kron(so,so))) @ np.kron(so,np.kron(CNOT,so)) @ ψ

O = 2*np.sqrt(2)*np.kron(sx,np.kron(sz,np.kron(so,sx)))
O = O - 2*np.sqrt(2)*np.kron(sy,np.kron(sz,np.kron(sz,sy)))
O = O + 2*np.sqrt(2)*np.kron(sz,np.kron(so,np.kron(sz,sz)))

print(ψp @ O @ ψp)

5.656854249492378


As before, users can change the initial state. We were unable to find a simple combination (or any combination) which produced $\Gamma_{r} > 4\sqrt{2} \approx 5.657$

## Version information:

In [26]:
import qiskit_ibm_runtime

qiskit_ibm_runtime.version.get_version_info()

'0.25.0'

In [27]:
import qiskit 

qiskit.version.get_version_info()

'1.1.1'